### PR TITLE
Port apply-geolocation-rules to augur curate

### LINKED ISSUE #1003: augur curate apply-geolocation-rules
### Tasks

- [x] https://github.com/nextstrain/augur/pull/1491
- [ ] Release new version of Augur
- [ ] Update [usages](https://github.com/search?q=org%3Anextstrain+vendored%2Fapply-geolocation-rules&type=code) of `vendored/apply-geolocation-rules` to use `augur curate apply-geolocation-rules`
- [ ] Update [creating-an-ingest-workflow docs](https://github.com/nextstrain/docs.nextstrain.org/blob/master/src/tutorials/creating-a-pathogen-repo/creating-an-ingest-workflow.rst) to point to `augur curate apply-geolocation-rules`
- [ ] Remove `apply-geolocation-rules` from nextstrain/ingest

### Context
We have a great list of [manually curated geolocation rules in ncov-ingest](https://github.com/nextstrain/ncov-ingest/blob/d5703c4af88f5d1532f9065422e1764ee7d47875/source-data/gisaid_geoLocationRules.tsv) that would be helpful to be used as standard geolocation rules that can be applied to other pathogen data curation efforts.

#### How the geolocation rules work

Instead of keeping track of manual annotations for locations on a per record/strain basis, these geolocation rules can be applied across all records. The geolocation rules TSV follows this format:
```
region/country/division/location<\t>region/country/division/location
```
The first set of locations are the expected "raw" locations that are in the metadata and the second set of locations after the tab are the "annotated" locations that will be applied to the metadata. Each geo resolution (`region`, `country`, `division`, `location`) is expected to be a column in the metadata. By using the `region/country/division/location` hierarchy, we ensure that locations with the same name are treated differently based on their full hierarchy.  If there are rules that can be applied across multiple locations, then a wildcard (`*`) can be used instead of a specific value. 

This might make more sense by walking through a specific example. 
Let's say you have the following locations in your metadata:

| region | country | division | location |
|--------|--------|----------|----------|
| North America | United States | New York | Buffalo |
| North America | United States | New York | New York |

You want to do the following:
- Change the location "New York" to "New York City"
- Change the division "New York" to "New York State"
- Change the country "United States" to "USA"

You can achieve this with these geolocation rules:
```
North America/United States/New York/New York<\t>North America/United States/New York/New York City
North America/United States/New York/*<\t>North America/United States/New York State/*
North America/United States/*/*<\t>North America/USA/*/*
```

The first rule will looks for the specific hierarchy to correct the location to "New York City". 
The second rule has a wildcard as the location, so it will correct all applicable divisions to "New York State". 
The third rule has wildcards for both division and location, so it will correct all applicable countries to "USA". 

### Description
`augur curate apply-geolocation-rules` can start from the [apply-geolocation-rules script from ingest](https://github.com/nextstrain/ingest/blob/main/apply-geolocation-rules).

### PR DESCRIPTION
**Pull request by victorlin**
## Description of proposed changes

Porting over from nextstrain/ingest with minimal functional changes.

## Related issue(s)

- Part of #1003
- Tests first written in https://github.com/nextstrain/mpox/pull/54
- Tests later moved to https://github.com/nextstrain/ingest/pull/26

## Checklist

- [x] Tests ported from old PRs
- [x] Other tests added
- [x] Checks pass
- [x] If making user-facing changes, add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) summarizing the changes in this PR

### CODE REVIEW

**Review by joverlee521 — APPROVED** [review]
Thank you for picking this up @victorlin! I added another test in https://github.com/nextstrain/augur/pull/1491/commits/b80bc4e11452a7d7a1d01c836a98c24c5ab263e0 to make sure I'm not crazy for remembering that we can override default rules.

File: augur/curate/apply_geolocation_rules.py
Comment (victorlin) [inline augur/curate/apply_geolocation_rules.py:1]: I wanted to check test coverage but noticed codecov is failing due to #1111.

I'll try to fix it now but if it turns out to be too much work I'll just merge this PR today.

Comment (genehack) [reply augur/curate/apply_geolocation_rules.py:1]: silly question: why not just run `pytest-cov` locally?

File: augur/curate/apply_geolocation_rules.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+from augur.errors import AugurError
+from augur.io.print import print_err
+from augur.utils import first_line
+
+
+class CyclicGeolocationRulesError(AugurError):
+    pass
+
+
+def load_geolocation_rules(geolocation_rules_file):
```
Comment (genehack) [inline augur/curate/apply_geolocation_rules.py:14]: silly question #2: do we eschew types in `augur`?

File: augur/curate/apply_geolocation_rules.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            countries: {
+                divisions: {
+                    locations: corrected_geolocations_tuple
+                }
+            }
+        }
+    }
+    """
+    geolocation_rules = defaultdict(lambda: defaultdict(lambda: defaultdict(dict)))
+    with open(geolocation_rules_file, 'r') as rules_fh:
```
Comment (genehack) [inline augur/curate/apply_geolocation_rules.py:29]: i'm gonna have a bunch of silly questions i think: is there a pattern in augur around who is responsible for handling (or, i guess, not handling) exceptions?

**Review by genehack — APPROVED** [review]

File: augur/curate/apply_geolocation_rules.py
Comment (victorlin) [reply augur/curate/apply_geolocation_rules.py:1]: Codecov has a nice interface to show coverage changes and it's something that's more visible to everyone looking at the PR (also, I guess I'm kinda lazy)

File: augur/curate/apply_geolocation_rules.py (hunk 1 above)
Comment (victorlin) [reply augur/curate/apply_geolocation_rules.py:14]: We don't, you'll find some here and there in two forms:

1. Types in docstrings

    https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/augur/io/metadata.py#L33-L39

2. Type annotations

    https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/augur/filter/subsample.py#L383-L386

These are picked up and rendered in docs such as [`augur.io.read_metadata`](https://docs.nextstrain.org/projects/augur/en/stable/api/public/augur.io.html#augur.io.read_metadata):

augur.io.read_metadata

But typing isn't ubiquitous and what's typed [isn't even validated](https://github.com/nextstrain/augur/blob/295f2806ab2a2ff71a692bdb7a56ede82e728caa/pyrightconfig.json#L16).

Historically most of the codebase has been untyped and we as a team have only recently begun using types in untyped languages. It hasn't been a priority to go back and address, but that's something I would love to see happen.

File: augur/curate/apply_geolocation_rules.py (hunk 2 above)
Comment (victorlin) [reply augur/curate/apply_geolocation_rules.py:29]: Yes, see #642.

Keep the questions coming! It's good to know what seems weird from a fresh perspective. Also this comment prompted me go back and update that outdated issue.

File: augur/curate/apply_geolocation_rules.py
Comment (victorlin) [reply augur/curate/apply_geolocation_rules.py:1]: Codecov is back: https://github.com/nextstrain/augur/pull/1491#issuecomment-2192736377

File: augur/curate/apply_geolocation_rules.py (hunk 1 above)
Comment (genehack) [reply augur/curate/apply_geolocation_rules.py:14]: > most of the codebase has been untyped and we as a team have only recently begun using types in untyped languages. It hasn't been a priority to go back and add more types to existing code, but that's something I would love to see happen.

one way to expand the use of types would be to have an general expectation that any new code being added would (at a minimum) annotate method signatures...

**Pull request merged**